src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
+        /// </remarks>
+        /// <exception cref="FormatException">Thrown if the given format isn't supported.</exception>
         /// <returns>True if successfully parsed.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value)
+        public static unsafe bool TryParse(ref this BufferReader<byte> reader, out int value, char standardFormat = '\0')
         {
             ReadOnlySpan<byte> unread = reader.UnreadSpan;
-            if (Utf8Parser.TryParse(unread, out value, out int consumed) && consumed < unread.Length)
+            if (!Utf8Parser.TryParse(unread, out value, out int consumed, standardFormat))
+            {
+                return false;
+            }
+
+            if (consumed < unread.Length)
             {
+                // The parser found a digit it wouldn't consume so we have all valid data.
                 reader.Advance(consumed);
                 return true;
             }
 
-            return TryParseSlow(ref reader, out value);
+            // If we ate all of our unread there may be more valid digits
+            return TryParseSlow(ref reader, out value, standardFormat);
         }
 
-        private static unsafe bool TryParseSlow(ref BufferReader<byte> reader, out int value)
+        private static unsafe bool TryParseSlow(ref BufferReader<byte> reader, out int value, char standardFormat)
         {
-            const int MaxLength = 15;
-
-            ReadOnlySpan<byte> unread = reader.UnreadSpan;
+            // Note that ints can have leading zeros ("-00000000000000127") which makes predicting the max
+            // length difficult. Typically the int will be less than 15 characters ("-2,147,483,648") so we'll
+            // work in those increments.
+            int bufferSize = 15;
 
-            if (unread.Length > MaxLength)
+            // Try using the stack for scratch space.
+            do
             {
-                // Fast path had enough space but couldn't find valid data
-                value = default;
-                return false;
-            }
-
-            byte* buffer = stackalloc byte[MaxLength];
-            Span<byte> tempSpan = new Span<byte>(buffer, MaxLength);
-            if (Utf8Parser.TryParse(reader.PeekSlow(tempSpan), out value, out int consumed))
+                // stackalloc is separated as we can't represent that PeekSlow will not capture the span.
+                byte* stackBuffer = stackalloc byte[bufferSize];
+                Span<byte> buffer = new Span<byte>(stackBuffer, bufferSize);
+                ReadOnlySpan<byte> peekSpan = reader.PeekSlow(buffer);
+
+                if (!Utf8Parser.TryParse(peekSpan, out value, out int consumed, standardFormat))
+                {
+                    // Overflow
+                    return false;
+                }
+
+                if (consumed < peekSpan.Length || peekSpan.Length < buffer.Length)
```

Comment:
what if the int is at the end of the reader's buffer?

Reply:
That is what the `peekSpan.Length < buffer.Length` is checking. If we couldn't fill our requested peek, we're at the end.

Reply:
I was wondering what happens if we request 15 bytes, get 15 bytes, and then parsed exactly 15 bytes? I gather we just loop and then the code path you are referring to will be hit. But this, together with my comment in the outer TryParse, means that if I parse exactly 15 bytes int, I will not only hit the slow code path, but also loop here. If it's not avoidable, then maybe it's ok, but it does seem wasteful at the first glance.

Reply:
You'd have to have exactly 15 bytes left in the reader's remaining span **and** have them all be valid to hit this case. Normal cases will be < 15. Together I think that case should be really rare.

You can also come in with more than 15 bytes if you're matching the end of the span. Same sort of rareness. We could change behavior here to check for these cases to avoid extra work, but I would expect them to impact the perf of common case- which is 14 bytes crossing segments.